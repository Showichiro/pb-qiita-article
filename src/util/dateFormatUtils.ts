/**
 * @description Convert a Date object to a string in ISO 8601 format.
 * @param {Date} date The Date object to convert.
 * @returns {string} The ISO 8601 string representation of the Date object.
 * @example
 * const date = new Date();
 * const isoString = dateToDatetimeString(date);
 */
export const dateToDatetimeString = (date: Date): string => {
  return date.toISOString();
};

/**
 * @description Convert a datetime string to a date string.
 * @param {string} datetime The datetime string to convert.
 * @returns {string} The date string representation of the datetime string.
 * @example
 * const datetime = "2022-01-01T12:00:00.000Z";
 * const dateString = dateTimetoDateString(datetime);
 */
export const dateTimetoDateString = (datetime: string): string => {
  return new Date(datetime).toISOString().split("T")[0];
};
