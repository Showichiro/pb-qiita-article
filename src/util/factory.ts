import { schema } from "@/db";
import { renderer } from "@/renderer";
import { swaggerUI } from "@hono/swagger-ui";
import { OpenAPIHono } from "@hono/zod-openapi";
import { drizzle, DrizzleD1Database } from "drizzle-orm/d1";
import { cors } from "hono/cors";
import { logger } from "hono/logger";

export type Env = {
  Variables: {
    db: DrizzleD1Database<typeof schema>;
  };
  Bindings: {
    DB: D1Database;
  };
};

export const createHonoWithDBAndOpenAPI = () => {
  const app = new OpenAPIHono<Env>();
  // cors
  app.use(
    "*",
    cors({
      origin: ["http://localhost:5173"],
    })
  );
  // openapi
  app
    .doc("/doc", {
      openapi: "3.0.0",
      info: {
        version: "1.0.0",
        title: "PB Qiita API",
      },
      servers: [
        {
          url: "http://localhost:5173/",
          description: "local",
        },
        {
          url: "https://pb-qiita-articles.pages.dev/",
          description: "PRD",
        },
      ],
    })
    .use(logger())
    .use(renderer)
    .get(
      "/ui",
      swaggerUI({
        url: "/doc",
      })
    );
  app.use(async (c, next) => {
    const db = drizzle(c.env.DB, { schema, logger: true });
    c.set("db", db);
    await next();
  });
  return app;
};
