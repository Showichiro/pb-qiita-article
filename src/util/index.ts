export * from "./typeUtils";
export * from "./testUtils";
export * from "./dateFormatUtils";
export * from "./factory";
export * from "./dateParamsUtils";
