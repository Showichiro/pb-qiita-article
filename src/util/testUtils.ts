import { Hono } from "@/lib";
import { testClient } from "hono/testing";
import { HtmlEscapedString } from "hono/utils/html";

/**
 * @description
 * Render a component to a string.
 * @param component
 * @returns string
 */
export const renderer = async (
  component: HtmlEscapedString | Promise<HtmlEscapedString>
): Promise<{ status: number; text: string }> => {
  const app = new Hono().get("/test", (c) => c.html(component));
  const res = await testClient(app).test.$get();
  return { status: res.status, text: await res.text() };
};
