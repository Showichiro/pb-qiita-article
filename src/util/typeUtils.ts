/**
 * Nullable type
 * @param T - Type to make nullable
 * @returns Nullable type of T
 * @example
 * type NullablePerson = Nullable<Person>;
 */
export type Nullable<T> = { [P in keyof T]: T[P] | null };
