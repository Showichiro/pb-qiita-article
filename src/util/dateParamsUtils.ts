import { dateToDatetimeString } from "./dateFormatUtils";

export const processDateParam = (
  date: string | Date | null | undefined,
): string | null => {
  if (typeof date === "string") return date;
  if (date == null) return null;
  return dateToDatetimeString(date);
};
