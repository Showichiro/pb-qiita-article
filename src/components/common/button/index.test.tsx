import { renderer } from "@/util";
import { ToTopButton } from ".";

describe("ToTopButton", () => {
  it("returns button", async () => {
    const { text } = await renderer(<ToTopButton />);
    expect(text).toMatchSnapshot();
  });
});
