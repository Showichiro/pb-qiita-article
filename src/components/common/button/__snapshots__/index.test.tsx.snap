// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToTopButton > returns button 1`] = `"<a href="#" aria-label="go to the top" class="btn btn-circle btn-accent fixed bottom-2 right-2 z-10"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24"><path d="m3.293 11.293 1.414 1.414L11 6.414V20h2V6.414l6.293 6.293 1.414-1.414L12 2.586l-8.707 8.707z"></path></svg></a>"`;
